working/0546-patterns/patterns-feature-specification.md

```diff
@@ -1600,44 +1718,51 @@ To type check a pattern `p` being matched against a value of type `M`:
+*   The type of an expression in a guard clause is not assignable to `bool`.
 
 *   `p` is in an irrefutable context, it has a required type `T`, and `M` is not
     assignable to `T`. *Destructuring and variable patterns can only be used in
     declarations and assignments if we can statically tell that the
-    destructuring and variable binding won't fail.*
+    destructuring and variable binding won't fail to match (though it might
+    throw a runtime exception from implicit downcasts from `dynamic`).*
+
+### Pattern uses
 
-### Switch expression type
+It is a compile-time error if the expression in a guard clause in a switch case
+or if-case statement is not assignable to `bool`.
 
 The static type of a switch expression is the least upper bound of the static
 types of all of the case expressions.
 
 ### Variables and scope
 
 Patterns often exist to introduce new variable bindings. A "wildcard" identifier
-named `_` in a variable or cast pattern never introduces a binding.
-
+named `_` in a variable pattern never introduces a binding.
 The variables a patterns binds depend on what kind of pattern it is:
 
 *   **Logical-or**: Does not introduce variables but may contain subpatterns
-    that do. If it a compile-time error if the two subpatterns do not introduce
+    that do. It is a compile-time error if the two subpatterns do not introduce
     the same variables with the same names and types.
 
-*   **Logical-and**, **null-check**, **null-assert**, **parenthesized**,
-    **list**, **map**, **record**, or **extractor**: These do not introduce
-    variables themselves but may contain subpatterns that do.
+    **TODO: "Same name" is underspecified and also probably
+    [too conservative][2473].**
+
+*   **Logical-and**, **cast**, **null-check**, **null-assert**,
+    **parenthesized**, **list**, **map**, **record**, or **extractor**: These do
+    not introduce variables themselves but may contain subpatterns that do.
+
+*   **Relational** or **constant**: These do not introduce any variables.
 
-*   **Relational**, **literal**, or **constant**: These do not introduce any
-    variables.
+*   **Variable**: When not in an assignment context, introduces a variable whose
+    name is the pattern's identifier. In a declaration context, the variable is
+    final if the surrounding `patternDeclaration` has a `final` modifier. In a
+    matching context, the variable is final if the variable pattern is marked
+    `final` and is not otherwise.
 
-*   **Variable** or **cast**: May contain type argument patterns. Introduces a
-    variable whose name is the pattern's identifier. The variable is final if
-    the surrounding `patternDeclaration` has a `final` modifier. *Variables in
-    switch cases don't occur inside pattern variable declarations and thus are
-    not final.*
+[2473]: https://github.com/dart-lang/language/issues/2473
 
-The scope where a pattern's variables are declared depends on the construct
-that contains the pattern:
```

Comment:
The scope where a local variable is declared, and where it's available for use, are not the same thing in the current specification. We need to define both.